README.md
# Http-Staror

#### With Typescript

### Http

```typescript
import { Err, Http, Status } from "http-staror"
```

```typescript
Http.setStatus("Accepted") // {status: 'Accepted',statusCode: 202,message: 'Accepted'}
```

```typescript
Http.setStatus("Accepted").setStatusCode(500) // { status: 'Accepted', statusCode: 500, message: 'Accepted' }

// Note: If you don't set status code explicitly then the status code would be inferred from the status you set.

//Most of the time would not want to set the status code explicitly unless you are setting a custom status that does not exist in the package
```

```typescript
Http.setStatus("Accepted").setStatusCode(500).setMessage("It is accepted") // { status: 'Accepted', statusCode: 500, message: 'It is accepted' }

// Note: If you don't set a message explicitly then the message would be the same as status
```

```typescript
Http.setStatus("x-hello").setStatusCode(303).setMessage("custom status") // { status: 'hello', statusCode: 303, message: 'custom status' }

// Note: To set a custom status you must prefix with "x-" also you have to set a status code otherwise it will be set to 200
```

### Err

```typescript
/** Supported Functions ⬇⬇
 * setProduction()
 * setStatus()
 * setFilePath()
 * setWhere()
 * setIsOperational()
 * setLineNumber()
 * setMessage()
 * setNoStack()
 * setStatusCode()
 * setUniqueIdentifier()
 */
```

```typescript
Err.setStatus("Ok").setMessage("custom status")

// Output ⬇
{
  filePath: null
  isOperational: true
  isProduction: false
  lineNumber: null
  message: "custom status""
  stack: "Error: custom status"\n    at _a2.setMessage (http://localhost:5173/node_modules/.vite/deps/http-staror.js?v=a66ceed3:146:39)\n    at http://localhost:5173/src/main.ts:2:39"
  status: "Ok"
  statusCode: 202
  uniqueIdentifier: null
  where: null
}

```

```typescript
Err.setProduction().setStatus("Ok").setMessage("Something Went Wrong!")

// Output ⬇
{
  filePath: null
  isOperational: true
  isProduction: true
  lineNumber: null
  message: "Something Went Wrong!"
  stack: null
  status: "Ok"
  statusCode: 200
  uniqueIdentifier: null
  where: null
}

// Note: Only set production for testing. This package already contains some to code to detect if it is in production or development mode.

// It will look at process.env.NODE_ENV to detect the mode

Err.setProduction(false).setStatus("Ok").setMessage("Something Went Wrong!")

// Warning: You can also setProduction to false therefore it will will be set to development mode even though your application is running in production mode.
```

```typescript
Err.setStatus("Ok").setMessage("Something Went Wrong!").setNoStack()

// Output ⬇
{
  filePath: null
  isOperational: true
  isProduction: true
  lineNumber: null
  message: "Something Went Wrong!"
  stack: null
  status: "Ok"
  statusCode: 200
  uniqueIdentifier: null
  where: null
}

// Note: You can set it to no stack mode therefore the stack will be null in development. In production stack is always null.
```

```typescript
Err.setStatus("Ok")
  .setMessage("Something Went Wrong!")
  .setNoStack()
  .setIsOperational(false)

// Output ⬇
{
  filePath: null
  isOperational: false
  isProduction: true
  lineNumber: null
  message: "Something Went Wrong!"
  stack: null
  status: "Ok"
  statusCode: 200
  uniqueIdentifier: null
  where: null
}

// Note: You can set is operational to false. By default it is true. If there is any unhandled error in the application that error won't have is operational property.
```

```typescript
// You can check if an error is instance your custom error
console.log(e.message instanceof Err) // boolean
```

```typescript
Err.setStatus("Ok")
  .setMessage("Something Went Wrong!")
  .setFilePath("src/app.ts")
  .setWhere("inside main function")
  .setLineNumber(24)
  .setUniqueIdentifier("axdf")
  .setNoStack()

// Output ⬇
{
  filePath: "src/app.ts"
  isOperational: true
  isProduction: true
  lineNumber: 24
  message: "Something Went Wrong!"
  stack: null
  status: "Ok"
  statusCode: 200
  uniqueIdentifier: "axdf"
  where: "inside main function"
}
```

```typescript
// you can throw error like this
try {
  //...do this
} catch (error) {
  throw Err.setStatus("InternalServerError").setWhere(
    "getVideosDurationString()"
  )
}

// This way structure of the error object is always same and predictable.
```

### Status

```typescript
console.log(Status.Ok)
// {value: "Ok", statusCode: 200}
```


library/CHANGELOG.md
# http-staror

## 1.0.7

### Patch Changes

- npm error solution 2

## 1.0.6

### Patch Changes

- npm error solution 1

## 1.0.5

### Patch Changes

- - spelling corrected - doc for throwing error

## 1.0.4

### Patch Changes

- Status doc add

## 1.0.3

### Patch Changes

- ga -test

## 1.0.2

### Patch Changes

- small update

## 1.0.1

### Patch Changes

- added readme


library/src/Http/Http.ts
const HttpStatuses = {
  Continue: { value: "Continue", code: 100 },
  SwitchingProtocols: { value: "SwitchingProtocols", code: 101 },
  Processing: { value: "Processing", code: 102 },
  EarlyHints: { value: "EarlyHints", code: 103 },
  Ok: { value: "Ok", code: 200 },
  Created: { value: "Created", code: 201 },
  Accepted: { value: "Accepted", code: 202 },
  NonAuthoritativeInformation: {
    value: "NonAuthoritativeInformation",
    code: 203,
  },
  NoContent: { value: "NoContent", code: 204 },
  ResetContent: { value: "ResetContent", code: 205 },
  PartialContent: { value: "PartialContent", code: 206 },
  MultiStatus: { value: "MultiStatus", code: 207 },
  AlreadyReported: { value: "AlreadyReported", code: 208 },
  IMUsed: { value: "IMUsed", code: 226 },
  MultipleChoices: { value: "MultipleChoices", code: 300 },
  MovedPermanently: { value: "MovedPermanently", code: 301 },
  Found: { value: "Found", code: 302 },
  SeeOther: { value: "SeeOther", code: 303 },
  NotModified: { value: "NotModified", code: 304 },
  UseProxy: { value: "UseProxy", code: 305 },
  TemporaryRedirect: { value: "TemporaryRedirect", code: 307 },
  PermanentRedirect: { value: "PermanentRedirect", code: 308 },
  BadRequest: { value: "BadRequest", code: 400 },
  Unauthorized: { value: "Unauthorized", code: 401 },
  PaymentRequired: { value: "PaymentRequired", code: 402 },
  Forbidden: { value: "Forbidden", code: 403 },
  NotFound: { value: "NotFound", code: 404 },
  MethodNotAllowed: { value: "MethodNotAllowed", code: 405 },
  NotAcceptable: { value: "NotAcceptable", code: 406 },
  ProxyAuthenticationRequired: {
    value: "ProxyAuthenticationRequired",
    code: 407,
  },
  RequestTimeout: { value: "RequestTimeout", code: 408 },
  Conflict: { value: "Conflict", code: 409 },
  Gone: { value: "Gone", code: 410 },
  LengthRequired: { value: "LengthRequired", code: 411 },
  PreconditionFailed: { value: "PreconditionFailed", code: 412 },
  PayloadTooLarge: { value: "PayloadTooLarge", code: 413 },
  URITooLong: { value: "URITooLong", code: 414 },
  UnsupportedMediaType: { value: "UnsupportedMediaType", code: 415 },
  RangeNotSatisfiable: { value: "RangeNotSatisfiable", code: 416 },
  ExpectationFailed: { value: "ExpectationFailed", code: 417 },
  ImATeapot: { value: "ImATeapot", code: 418 },
  MisdirectedRequest: { value: "MisdirectedRequest", code: 421 },
  UnprocessableEntity: { value: "UnprocessableEntity", code: 422 },
  Locked: { value: "Locked", code: 423 },
  FailedDependency: { value: "FailedDependency", code: 424 },
  TooEarly: { value: "TooEarly", code: 425 },
  UpgradeRequired: { value: "UpgradeRequired", code: 426 },
  PreconditionRequired: { value: "PreconditionRequired", code: 428 },
  TooManyRequests: { value: "TooManyRequests", code: 429 },
  RequestHeaderFieldsTooLarge: {
    value: "RequestHeaderFieldsTooLarge",
    code: 431,
  },
  UnavailableForLegalReasons: {
    value: "UnavailableForLegalReasons",
    code: 451,
  },
  InternalServerError: { value: "InternalServerError", code: 500 },
  NotImplemented: { value: "NotImplemented", code: 501 },
  BadGateway: { value: "BadGateway", code: 502 },
  ServiceUnavailable: { value: "ServiceUnavailable", code: 503 },
  GatewayTimeout: { value: "GatewayTimeout", code: 504 },
  HTTPVersionNotSupported: { value: "HTTPVersionNotSupported", code: 505 },
  VariantAlsoNegotiates: { value: "VariantAlsoNegotiates", code: 506 },
  InsufficientStorage: { value: "InsufficientStorage", code: 507 },
  LoopDetected: { value: "LoopDetected", code: 508 },
  NotExtended: { value: "NotExtended", code: 510 },
  NetworkAuthenticationRequired: {
    value: "NetworkAuthenticationRequired",
    code: 511,
  },
  UnparsableResponseHeaders: {
    value: "UnparsableResponseHeaders",
    code: 600,
  },
  ConnectionTimedOut: { value: "ConnectionTimedOut", code: 601 },
  RequestTimedOut: { value: "RequestTimedOut", code: 602 },
  NameNotResolved: { value: "NameNotResolved", code: 603 },
} as const

// single value object type on HttpStatuses
export interface IHttpStatus {
  value: keyof typeof HttpStatuses
  code: number
}

export const Status = HttpStatuses as {
  [key in keyof typeof HttpStatuses]: IHttpStatus
}

type HttpStatusKeysType = keyof typeof HttpStatuses

export type ExtendedHttpStatusKeysType = HttpStatusKeysType | `x-${string}`

export type setStatusParamType = ExtendedHttpStatusKeysType | IHttpStatus

export interface IStatus {
  status: ExtendedHttpStatusKeysType
  statusCode: number
  message: string
}

interface IStatusPrototype extends IStatus {
  readonly setStatusCode: (statusCode: number) => IStatusPrototype
  readonly setMessage: (message: string) => IStatusPrototype
}

class Http implements IStatusPrototype {
  declare readonly status: ExtendedHttpStatusKeysType
  declare statusCode: number
  declare message: string

  static readonly setStatus = (status: setStatusParamType) => {
    // if status param is an object then the value would be a string that exists inside of HttpStatuses
    if (status instanceof Object)
      status = status.value as ExtendedHttpStatusKeysType
    // remove x- if it is a custom status
    else if (status.startsWith("x-"))
      status = status.slice(2) as ExtendedHttpStatusKeysType

    const statusObj = Object.create(new Http()) as IStatusPrototype
    statusObj.status = status

    // if the status exists in HttpStatus then the status code would be relative to the key otherwise a generic 200
    statusObj.statusCode =
      status in HttpStatuses
        ? HttpStatuses[status as HttpStatusKeysType].code
        : 200

    statusObj.message = status

    return statusObj
  }

  setStatusCode(statusCode: number) {
    this.statusCode = statusCode
    return this
  }

  setMessage(message: string) {
    this.message = message
    return this
  }
}

export default Http
